scene/resources/2d/primitive_meshes_2d.cpp

```diff
@@ -0,0 +1,916 @@
+}
+
+AABB PrimitiveMesh2D::get_aabb() const {
+	if (pending_request) {
+		_update();
+	}
+
+	return aabb;
+}
+
+RID PrimitiveMesh2D::get_rid() const {
+	if (pending_request) {
+		_update();
+	}
+	return mesh;
+}
+
+void PrimitiveMesh2D::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("get_mesh_arrays"), &PrimitiveMesh2D::get_mesh_arrays);
+
+	ClassDB::bind_method(D_METHOD("set_antialiased", "p_antialiased"), &PrimitiveMesh2D::set_antialiased);
+	ClassDB::bind_method(D_METHOD("is_antialiased"), &PrimitiveMesh2D::is_antialiased);
+
+	ClassDB::bind_method(D_METHOD("set_custom_aabb", "aabb"), &PrimitiveMesh2D::set_custom_aabb);
+	ClassDB::bind_method(D_METHOD("get_custom_aabb"), &PrimitiveMesh2D::get_custom_aabb);
+
+	ClassDB::bind_method(D_METHOD("request_update"), &PrimitiveMesh2D::request_update);
+
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "antialiased"), "set_antialiased", "is_antialiased");
+	ADD_PROPERTY(PropertyInfo(Variant::AABB, "custom_aabb", PROPERTY_HINT_NONE, "suffix:px"), "set_custom_aabb", "get_custom_aabb");
+
+	GDVIRTUAL_BIND(_create_mesh_array);
+}
+
+void PrimitiveMesh2D::_validate_property(PropertyInfo &p_property) const {
+	// Remove unused lightmap size hint.
+	if (p_property.name == "lightmap_size_hint") {
+		p_property.usage = PROPERTY_USAGE_NONE;
+	}
+}
+
+Array PrimitiveMesh2D::get_mesh_arrays() const {
+	return surface_get_arrays(0);
+}
+
+bool PrimitiveMesh2D::is_antialiased() const {
+	return antialiased;
+}
+
+void PrimitiveMesh2D::set_antialiased(bool p_antialiased) {
+	if (antialiased == p_antialiased) {
+		return;
+	}
+	antialiased = p_antialiased;
+	request_update();
+}
+
+void PrimitiveMesh2D::set_custom_aabb(const AABB &p_custom) {
+	if (p_custom.is_equal_approx(custom_aabb)) {
+		return;
+	}
+	custom_aabb = p_custom;
+	RS::get_singleton()->mesh_set_custom_aabb(mesh, custom_aabb);
+	emit_changed();
+}
+
+AABB PrimitiveMesh2D::get_custom_aabb() const {
+	return custom_aabb;
+}
+
+PrimitiveMesh2D::PrimitiveMesh2D() {
+	ERR_FAIL_NULL(RenderingServer::get_singleton());
+	mesh = RenderingServer::get_singleton()->mesh_create();
+}
+
+PrimitiveMesh2D::~PrimitiveMesh2D() {
+	ERR_FAIL_NULL(RenderingServer::get_singleton());
+	RenderingServer::get_singleton()->free(mesh);
+}
+
+/**
+  RectangleMesh2D
+*/
+
+void RectangleMesh2D::_create_mesh_array(Array &p_arr) const {
+	int i, j, prevrow, thisrow, point;
+	float x, y;
+
+	Vector<Vector2> points;
+	Vector<Color> colors;
+	Vector<Vector2> uvs;
+	Vector<int> indices;
+	point = 0;
+
+	Vector2 rect_size = size;
+	Vector2 uv_scale = Vector2(1, 1);
+	Vector2 uv_offset = Vector2(0, 0);
+	if (antialiased) {
+		rect_size -= Vector2(FEATHER_SIZE, FEATHER_SIZE);
+		uv_scale = Vector2(rect_size.x / (size.x + FEATHER_SIZE), rect_size.y / (size.y + FEATHER_SIZE));
+		uv_offset = Vector2(FEATHER_SIZE / (size.x + FEATHER_SIZE), FEATHER_SIZE / (size.y + FEATHER_SIZE));
+	}
+	Vector2 start_pos = rect_size * -0.5;
```

Comment:
Regarding the uvs I'm not sure if it's desired that toggling antialiasing makes the texture "move":
![VtEHYD6Bnq](https://github.com/user-attachments/assets/f7884934-f185-4f88-a576-2b985eb733cb)

Shouldn't it be shown exactly the same within the original rect regardless of the antialiasing, plus for antialiasing just add additional "smears" around? :thinking: (I think this would need more vertices for the corners though so it would interpolate correctly)

Also is it supposed to be `FEATHER_SIZE`-wide on each side, or `FEATHER_SIZE * 0.5` on each side? Cause currently the original rect size is growing by `FEATHER_SIZE`, not by `2 * FEATHER_SIZE`.

Reply:
Right now turning on antialiasing shrinks each shape by `FEATHER_SIZE * 0.5` and adds a border of size `FEATHER_SIZE`, so that the halfway point between solid and transparent lines up with where the border used to be. I did it that way to preserve the visual size, which works well at sizes a bit bigger than your tiny 5px rectangle. Maybe we could scale the feather size with the regular size (for tiny shapes), but it would become a bit unintuitive.

I would have liked to keep the texture shown the same way when antialiasing is toggled, but since there is no general UV scale/offset available to manipulate I think the only way would be to give the non-antialliased shape a shrunk UV map, which would be awkward as well for people who just want to use normal a primitive shape.